Assets/Plugins/Source/Editor/EditorWindows/NewConfigWindow.cs

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2024 PlayEveryWare
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+// Uncomment the following line to see the experimental new config window
+//#define ENABLE_NEW_CONFIG_WINDOW
+
+#if ENABLE_NEW_CONFIG_WINDOW
+namespace PlayEveryWare.EpicOnlineServices.Editor.Windows
+{
+    using System;
+    using System.Threading.Tasks;
+    using UnityEditor;
+    using UnityEngine;
+
+    /// <summary>
+    /// Creates the view for showing the eos plugin editor config values.
+    /// </summary>
+    [Serializable]
+    public class NewConfigWindow : EOSEditorWindow
+    {
+        /// <summary>
+        /// The editor for the product information that is shared across all
+        /// platforms (represents information that is common to all
+        /// circumstances).
+        /// </summary>
+        private ConfigEditor<ProductConfig> _productConfigEditor;
+
+        public NewConfigWindow() : base("EOS Configuration") { }
+
+        [MenuItem("EOS Plugin/[Experimental] New Config")]
+        public static void ShowWindow()
+        {
+            var window = GetWindow<NewConfigWindow>();
+            window.SetIsEmbedded(false);
+        }
+
+        protected override async Task AsyncSetup()
+        {
+            _productConfigEditor = new();
+            await _productConfigEditor.LoadAsync();
+        }
+
+        protected override void RenderWindow()
+        {
+            _ = _productConfigEditor.RenderAsync();
```

Comment:
There's something weird with this, "synchronously open a UI, asynchronously load stuff, take actions before done", kind of approach... but I don't think Unity's UI system really gives you a much better way to do things. Here this is fine, I don't strongly feel it should change.

Some part of me wonders if we should use "ContinueWith" to set some kind of, "okay this screen is now ready for you to do things" state, and before that's finished the user can't take actions.

Reply:
Unity's UI system (used for rendering things in scenes) is distinct from Unity's UI Toolkit (used for rendering things in the editor). These functions take place in the context of the latter. The two technologies have very different origins and hence different design paradigms.

The distinction between these different systems is documented by Unity [here](https://docs.unity3d.com/Manual/UIToolkits.html).

Reply:
Ah, I may have over-used a word. I'm unconcerned with what happens here at runtime, just editor time. It's *concerning to me* that at editor time, the developer implementing our plugin *could theoretically* run actions that the editor window is not yet ready to receive because asynchronous operations are running. I wish we could design our editor tools in a way that this is not possible.

However, several of our windows already work like this. I'm not greatly concerned, and I also don't have a solution for what to do for the UX. Maybe in years from now where we're doing all sorts of even the smallest UX improvements, my idea here could be implemented. I'm not gonna block this PR on the idea.